instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/HostIdResourceProvider.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.resources;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.ConditionalResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.ResourceAttributes;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * {@link ResourceProvider} for automatically configuring <code>host.id</code> according to <a
+ * href="https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/host.md#non-privileged-machine-id-lookup">the
+ * semantic conventions</a>
+ */
+public final class HostIdResourceProvider implements ConditionalResourceProvider {
+
+  private static final Logger logger = Logger.getLogger(HostIdResourceProvider.class.getName());
+
+  public static final String REGISTRY_QUERY =
+      "reg query HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Cryptography /v MachineGuid";
+
+  private final Supplier<OsType> getOsType;
+
+  private final Function<Path, List<String>> machineIdReader;
+
+  private final Supplier<ExecResult> queryWindowsRegistry;
+
+  enum OsType {
+    WINDOWS,
+    LINUX,
+    OTHER
+  }
+
+  static class ExecResult {
+    int exitCode;
+    List<String> lines;
+
+    public ExecResult(int exitCode, List<String> lines) {
+      this.exitCode = exitCode;
+      this.lines = lines;
+    }
+  }
+
+  public HostIdResourceProvider() {
+    this(
+        HostIdResourceProvider::getOsType,
+        HostIdResourceProvider::readMachineIdFile,
+        HostIdResourceProvider::queryWindowsRegistry);
+  }
+
+  // Visible for testing
+
+  HostIdResourceProvider(
+      Supplier<OsType> getOsType,
+      Function<Path, List<String>> machineIdReader,
+      Supplier<ExecResult> queryWindowsRegistry) {
+    this.getOsType = getOsType;
+    this.machineIdReader = machineIdReader;
+    this.queryWindowsRegistry = queryWindowsRegistry;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    OsType osType = getOsType.get();
+    switch (osType) {
+      case WINDOWS:
+        return readWindowsGuid();
+      case LINUX:
+        return readLinuxMachineId();
+      case OTHER:
+        break;
+    }
+    logger.info("Unsupported OS type: " + osType);
```

Comment:
This isn't a good idea. On mac this would always print `Unsupported OS type: OTHER` that is not particularly useful. I'd move the `return Resource.empty();` inside the `OTHER` case of reduce log level.

Reply:
I think I'll just reduce the log level to `fine` - so you can still get an idea why mac doesn't detect anything.